test/vegeta/pacers/combined_pacer.go

```diff
@@ -0,0 +1,120 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package pacers
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+	"time"
+
+	vegeta "github.com/tsenart/vegeta/lib"
+)
+
+// combinedPacer is a Pacer that combines multiple Pacers and runs them sequentially when being used for attack.
+type combinedPacer struct {
+	// pacers is a list of pacers that will be used sequentially for attack, must be more than 1 pacer.
+	pacers []vegeta.Pacer
+	// durations is the list of durations for the given Pacers, must have the same length as Pacers.
+	durations []time.Duration
+
+	// totalDuration is sum of the given Durations.
+	totalDuration uint64
+	// stepDurations is the accumulative duration of each step calculated from the given Durations.
+	stepDurations []uint64
+
+	curtPacerIndex  uint
+	prevElapsedHits uint64
+	prevElapsedTime uint64
+}
+
+// NewCombined returns a new CombinedPacer with the given config.
+func NewCombined(pacers []vegeta.Pacer, durations []time.Duration) (vegeta.Pacer, error) {
+	if len(pacers) == 0 || len(durations) == 0 || len(pacers) != len(durations) || len(pacers) == 1 {
+		return nil, errors.New("configuration for this CombinedPacer is invalid")
+	}
+
+	var totalDuration uint64
+	stepDurations := make([]uint64, len(pacers))
+	for i, duration := range durations {
+		totalDuration += uint64(duration)
+		if i == 0 {
+			stepDurations[i] = uint64(duration)
+		} else {
+			stepDurations[i] = stepDurations[i-1] + uint64(duration)
+		}
+	}
+	pacer := &combinedPacer{
+		pacers:    pacers,
+		durations: durations,
+
+		totalDuration: totalDuration,
+		stepDurations: stepDurations,
+
+		curtPacerIndex:  0,
+		prevElapsedHits: 0,
+		prevElapsedTime: 0,
+	}
+	return pacer, nil
+}
+
+// combinedPacer satisfies the Pacer interface.
+var _ vegeta.Pacer = &combinedPacer{}
+
+// String returns a pretty-printed description of the combinedPacer's behaviour.
+func (cp *combinedPacer) String() string {
+	var sb strings.Builder
+	for i := range cp.pacers {
+		pacerStr := fmt.Sprintf("Pacer: %s, Duration: %s\n", cp.pacers[i], cp.durations[i])
+		sb.WriteString(pacerStr)
+	}
+	return sb.String()
+}
+
+func (cp *combinedPacer) Pace(elapsedTime time.Duration, elapsedHits uint64) (time.Duration, bool) {
+	pacerTimeOffset := uint64(elapsedTime) % cp.totalDuration
+	pacerIndex := cp.getPacerIndex(pacerTimeOffset)
+	// The pacer must be the same or the next neighbor of the current pacer, otherwise stop the attack.
+	if pacerIndex != cp.curtPacerIndex &&
```

Comment:
Can we just short-circuit to `if pi == len(cp.pacers)` ?
I.e. I don't really see how it can be not current and not the next one (unless the duration is so short that we jumped over it).

Reply:
After adding the `> 1 * time.Second` restriction, there will be very little chance to jump over the pacer, unless we intend to break it. I have removed all the checks here.